src/main/java/org/broadinstitute/hellbender/tools/spark/sv/ISZMetricsCollectorSpark.java

```diff
@@ -0,0 +1,150 @@
+package org.broadinstitute.hellbender.tools.spark.sv;
+
+import htsjdk.samtools.SAMReadGroupRecord;
+import htsjdk.samtools.SamPairUtil;
+import htsjdk.samtools.metrics.MetricsFile;
+import org.apache.spark.api.java.JavaRDD;
+import org.broadinstitute.hellbender.metrics.MetricAccumulationLevel;
+import org.broadinstitute.hellbender.tools.picard.analysis.InsertSizeMetrics;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+
+/**
+ * Worker class to collect insert size metrics.
+ */
+public class ISZMetricsCollectorSpark implements Serializable{
+    private static final long serialVersionUID = 1L;
+
+    // uses TreeMap because it is serializable and sorted
+    private TreeMap<Integer, Long> hist;
+
+    public ISZMetricsCollectorSpark(final JavaRDD<GATKRead> filteredReads,
+                                    final Set<MetricAccumulationLevel> accumLevels,
+                                    final List<SAMReadGroupRecord> samRgRecords,
+                                    final double devTolerance) {
+
+        final Map<Integer, Long> toBeCasted = filteredReads.map(read -> Math.abs(read.getFragmentLength()) ).countByValue();
+        hist = new TreeMap<>(toBeCasted);
+    }
+
+    public void produceMetricsFile(MetricsFile<InsertSizeMetrics, ?> metricsFile) {
+
+        InsertSizeMetrics metric = new InsertSizeMetrics();
+        metric.PAIR_ORIENTATION = SamPairUtil.PairOrientation.FR; // what does this mean?
+
+        metric.MIN_INSERT_SIZE = hist.firstKey();
+        metric.MAX_INSERT_SIZE = hist.lastKey();
+
+        collectMedianAndMAD(metric);
+
+        collectPercentages(metric);
+
+        metricsFile.addMetric(metric);
+    }
+
+    private void collectMedianAndMAD(InsertSizeMetrics metric){
```

Comment:
this whole method needs to be massively simplified using libraries - see ideas below. At the very least separate a generic computation of median to a separate method and do the same to MAD.